### PR TITLE
Ninja support and new build_system setting

### PR DESCRIPTION
**Pull request by ethteck**

### CODE REVIEW

File: import.py [hunk 1]
```diff
@@ -540,16 +552,27 @@ def main() -> None:
     )
     args = parser.parse_args()
 
-    make_flags = args.make_flags + ["PERMUTER=1"]
+    settings: Mapping[str, object] = {}
+    settings_dir = find_root_dir(args.c_file, SETTINGS_FILES)
```
Comment (simonlindholm) [inline import.py:556]: Maybe add `["Makefile", "makefile", "build.ninja"]` to the list of searched files, and remove the find_root_dir calls in find_build_command_line? Feels a bit convoluted to search for root dir twice.

File: import.py [hunk 2]
```diff
@@ -111,19 +111,16 @@ def create_directory(func_name: str) -> str:
             pass
 
 
-def find_makefile_dir(filename: str) -> str:
+def find_root_dir(filename: str, pattern: List[str]) -> str:
```
Comment (simonlindholm) [inline import.py:114]: Optional[str] (can you run mypy?)

File: import.py [hunk 3]
```diff
@@ -111,19 +111,16 @@ def create_directory(func_name: str) -> str:
...
     while dirname and (not old_dirname or len(dirname) < len(old_dirname)):
-        for fname in ["makefile", "Makefile"]:
+        for fname in pattern:
             if os.path.isfile(os.path.join(dirname, fname)):
                 return dirname
         old_dirname = dirname
         dirname = os.path.dirname(dirname)
 
-    print(f"Missing makefile for file {filename}!", file=sys.stderr)
-    sys.exit(1)
```
Comment (simonlindholm) [inline import.py:125]: `return None` (or reinstate the error in this function, either way)

File: import.py [hunk 4]
```diff
@@ -162,16 +159,23 @@ def fixup_build_command(
 
 
 def find_build_command_line(
-    c_file: str, make_flags: List[str]
+    root_dir: str, c_file: str, make_flags: List[str], build_system: str
 ) -> Tuple[List[str], List[str], str]:
```
Comment (simonlindholm) [inline import.py:163]:
```suggestion
) -> Tuple[List[str], List[str]]:
```

File: import.py [hunk 5]
```diff
@@ -204,7 +217,7 @@ def find_build_command_line(
             else ""
         )
         print(
-            "Failed to find compile command from makefile output. "
+            "Failed to find compile command from build script output. "
             f"Please ensure 'make -Bn --debug=j {formatcmd(make_flags)}' "
```
Comment (simonlindholm) [inline import.py:221]: this will be misleading with ninja, maybe worth fixing

File: import.py [hunk 6]
```diff
@@ -215,28 +228,21 @@ def find_build_command_line(
...
-    return output[0], assembler, makefile_dir
+    return output[0], assembler
 
 
 PreserveMacros = Tuple[Pattern[str], Callable[[str], str]]
 
 
 def build_preserve_macros(
-    cwd: str, preserve_regex: Optional[str]
+    cwd: str, preserve_regex: Optional[str], data: Mapping[str, object]
```
Comment (simonlindholm) [inline import.py:244]: data -> settings

**Comment by nanaian** [comment]
Resolved the review comments, and `mypy` now outputs only the two errors that already exist on `master`.

File: import.py [hunk 7]
```diff
@@ -227,9 +230,9 @@ def find_build_command_line(
     if len(output) > 1:
         output_lines = "\n".join(map(formatcmd, output))
         print(
-            f"Error: found multiple compile commands for {rel_c_file}:\n{output_lines}\n"
-            "Please modify the build script such that if PERMUTER = 1, "
-            "only a single compile command is included.",
+            f"Error: found multiple compile commands for {rel_c_file}:\n{output_lines}\n{close_extra}"
```
Comment (simonlindholm) [inline import.py:233]:
```suggestion
            f"Error: found multiple compile commands for {rel_c_file}:\n{output_lines}\n"
```
(close_extra isn't defined here, nor would its definition from the if block above make sense here)

**Review by simonlindholm — APPROVED** [review]

**Pull request merged**